### PR TITLE
MDEV-24949: Enabling idle flushing (possible regression from MDEV-23855)

### PR DESCRIPTION
**Pull request by mysqlonarm**
- Currently page cleaner thread will stop flushing if
  dirty_pct < srv_max_dirty_pages_pct_lwm.

- If the server is not performing any activity then said resources/time
  could be used to flush the pending dirty pages and keep buffer pool
  clean for the next burst of the cycle. This flushing is called idle flushing.

- flushing logic underwent a complete revamp in 10.5.7/8
  and as part of the revamp idle flushing logic got removed.

- New proposed logic of idle flushing is based on updated logic of the
  page cleaner that will enable idle flushing if
  - buf page cleaner is idle
  - there are dirty pages (< srv_max_dirty_pages_pct_lwm)
  - server is not performing any activity

  Logic will kickstart the idle flushing bounded by srv_io_capacity.

(Thanks to Marko Makela for reviewing the patch and idea
 right from the its inception).

### CODE REVIEW

File: storage/innobase/buf/buf0flu.cc [hunk 1]
```diff
@@ -2095,9 +2120,11 @@ static os_thread_ret_t DECLARE_THREAD(buf_flush_page_cleaner)(void*)
...
       break;
 
-    if (buf_pool.page_cleaner_idle())
-      my_cond_wait(&buf_pool.do_flush_list,
-                   &buf_pool.flush_list_mutex.m_mutex);
+    if (buf_pool.page_cleaner_idle()
+	&& ((srv_max_dirty_pages_pct_lwm != 0
+             && UT_LIST_GET_LEN(buf_pool.flush_list) == 0)
+            || (srv_max_dirty_pages_pct_lwm == 0)))
+      my_cond_wait(&buf_pool.do_flush_list, &buf_pool.flush_list_mutex.m_mutex);
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:2127]: I think that your expression could be simpilfied to the following:
```c++

    if (buf_pool.page_cleaner_idle() &&
        (!srv_max_dirty_pages_pct_lwm ||
         !UT_LIST_GET_LEN(buf_pool.flush_list)))
```
(Note: We do not follow Heikki Tuuri’s style of splitting a line before an operator, but the common style of splitting after an operator.)
This condition may be causing unnecessary wake-ups in the case that `innodb_max_dirty_pages_pct_lwm` has been set but the `dirty_pct` does not exceed it. But, it would not be easy to move the `dirty_pct` computation earlier. To compensate for this, I think that we can add a `my_cond_wait()` call before the `continue` in the following:
```c++
unemployed:
      buf_pool.page_cleaner_set_idle(true);
      continue;
```

File: storage/innobase/buf/buf0flu.cc [hunk 2]
```diff
@@ -2220,6 +2258,9 @@ static os_thread_ret_t DECLARE_THREAD(buf_flush_page_cleaner)(void*)
 next:
 #endif /* !DBUG_OFF */
     mysql_mutex_lock(&buf_pool.flush_list_mutex);
+
+    if (idle_flush)
+      buf_pool.page_cleaner_set_idle(true);
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:2263]: Is this change really needed? If yes, why? I would expect `goto unemployed` to do the right thing. Note: during the idle flushing, we may write multiple batches of `srv_io_capacity` (`innodb_io_capacity`) pages in this loop.

File: storage/innobase/buf/buf0flu.cc [hunk 3]
```diff
@@ -139,6 +139,30 @@ inline void buf_pool_t::page_cleaner_wakeup()
   {
     page_cleaner_is_idle= false;
     pthread_cond_signal(&do_flush_list);
+  } else if (pct_lwm != 0.0 && dirty_pct != 0) {
+    /* dirty_pct < srv_max_dirty_pages_pct_lwm so it could be idle flushing
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:143]: We are trying to move to the common formatting style, where lines are split before `{` or after `}`. An exception is the opening brace after `switch (...) {`.

I would prefer to refer to user-visible parameters in comments or commit messages. It is a bit unfortunate that we have this `srv_` obfuscation layer in the internal variable names. So, I would prefer to write `innodb_max_dirty_pages_pct_lwm` in the comment.

File: storage/innobase/buf/buf0flu.cc [hunk 4]
```diff
@@ -139,6 +139,30 @@ inline void buf_pool_t::page_cleaner_wakeup()
...
+    use-case. signal page cleaner only if server is not active.
+
+    Why is last_activity_count not updated always?
+    - to understand this let's first understand when it server activity count
+      updated.
+    - it is updated on trx_commit and not on adding the page to the flush list.
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:149]: The former function `trx_commit()` is called `trx_t::commit()` in this branch.

**Review by dr-m — COMMENTED** [review]
Thank you. This looks mostly good. I am mainly wondering why the page cleaner loop is not using `buf_pool.last_activity_count` but a local variable of its own.

**Review by mysqlonarm — COMMENTED** [review]
last_activity_count in page cleaner is updated every time the idle flushing is evaluated (that is once per second). if the server becomes active during the said time then idle flushing is disabled. last_activity_count (tracked in buf_pool_t) is just used to avoid duplicate signal from the thread if initial signal is not yet processed by the page-cleaner.

**Review by dr-m — COMMENTED** [review]
I tested with the following:
```sql
--source include/have_innodb.inc
--source include/have_sequence.inc

create table t (a int primary key, c char(255) not null)
engine=innodb row_format=redundant default charset utf8;
insert into t select seq,'' from seq_1_to_1000;

sleep 3600;

drop table t;
```
invoked as
```sh
./mtr --mysqld=--innodb-page-size=4k --manual-gdb --mysqld=--innodb-max-dirty-pages-pct-lwm=5 innodb.f
```
I set a breakpoint on the assignment
```c++
    lsn_limit= buf_flush_sync_lsn;
```
and observed that this breakpoint would be hit about once per second, until `buf_pool.flush_list.count` reached 0 (at which point we will enter the unbounded wait). With this test, I confirmed that your change works.

I pushed a simpler version of this in 7ecee0c35af8085a8730fce9f2f5917c65406bb0, limiting all changes to `buf_flush_page_cleaner`. I hope that it would works for both you and @hgxl64. I only ran it in a debugger, using the above test case.

File: storage/innobase/buf/buf0flu.cc (hunk 2 above)
Comment (dr-m) [reply storage/innobase/buf/buf0flu.cc:2269]: Thank you. I now realized that if we fail to call this, then at the start of the next loop we would invoke `my_cond_timedwait()` and cause an unnecessary 1-second wakeup.

File: storage/innobase/buf/buf0flu.cc (hunk 1 above)
Comment (dr-m) [reply storage/innobase/buf/buf0flu.cc]: I verified that my suggestion to add the `my_cond_wait()` call does not work.

File: storage/innobase/buf/buf0flu.cc [hunk 5]
```diff
@@ -140,6 +140,35 @@ inline void buf_pool_t::page_cleaner_wakeup()
     page_cleaner_is_idle= false;
     pthread_cond_signal(&do_flush_list);
   }
+  else if (pct_lwm != 0.0 && dirty_pct != 0)
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:143]: By definition, `dirty_pct` is always nonzero here. We do not need that condition. Because this function will be called while holding `buf_pool.flush_list_mutex` and right after adding something to `buf_pool.flush_list`, there will be some work to do, and we only need the condition
```c++
else if (pct_lwm != 0.0)
```

File: storage/innobase/buf/buf0flu.cc [hunk 6]
```diff
@@ -140,6 +140,35 @@ inline void buf_pool_t::page_cleaner_wakeup()
...
+    How to avoid this?
+    - by allowing last_activity_count to updated when page-cleaner is made
+      active and has work to do. This ensures that the last_activity signal
+      is consumed by the page-cleaner before the next one is generated. */
+    ulint cached_activity_count= last_activity_count;
+    if (!srv_check_activity(&cached_activity_count))
+    {
+      page_cleaner_is_idle= false;
+      pthread_cond_signal(&do_flush_list);
+    }
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:170]: Please use `goto` to the identical code earlier in the function.

File: storage/innobase/buf/buf0flu.cc [hunk 7]
```diff
@@ -2135,10 +2168,19 @@ static os_thread_ret_t DECLARE_THREAD(buf_flush_page_cleaner)(void*)
...
       double(UT_LIST_GET_LEN(buf_pool.LRU) + UT_LIST_GET_LEN(buf_pool.free));
 
+    /* idle_flush steps kicks in
+    - if page cleaner is idle
+    - there are dirty pages (dirty_pct < innodb_max_dirty_pages_pct_lwm)
+    - server is idle (no trx_t::commit() activity)
+    - there are no pending reads (seen with io-bound workload) */
+    const bool idle_flush= buf_pool.page_cleaner_idle() &&
+      dirty_blocks != 0 && !srv_check_activity(&last_activity_count) &&
+      (buf_pool.n_pend_reads == 0);
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:2178]: The `dirty_blocks != 0 &&` is redundant here, because we just checked for `!dirty_blocks` a couple of statements earlier.

I think that we are computing this unnecessarily if `srv_max_dirty_pages_pct_lwm` is at its default value 0. The following should avoid that:
```c++
    bool idle_flush= false;

    if (lsn_limit);
    else if (srv_max_dirty_pages_pct_lwm != 0.0)
    {
      const ulint activity_count= srv_get_activity_count();
      if (activity_count != last_activity_count)
        last_activity_count= activity_count;
      else if (buf_pool.page_cleaner_idle() && buf_pool.n_pending_reads == 0)
      {
        idle_flush= true;
        buf_pool.update_last_activity_count(activity_count);
      }
      if (!idle_flush && dirty_pct < srv_max_dirty_pages_pct_lwm)
        goto unemployed;
    }
```
Note: This includes `buf_pool.update_last_activity_count()`, and I would pass the count as a parameter, to avoid a redundant call.

File: storage/innobase/buf/buf0flu.cc [hunk 8]
```diff
@@ -2152,6 +2194,8 @@ static os_thread_ret_t DECLARE_THREAD(buf_flush_page_cleaner)(void*)
       buf_flush_sync_lsn= 0;
 
     buf_pool.page_cleaner_set_idle(false);
+    if (idle_flush)
+      buf_pool.update_last_activity_count();
```
Comment (dr-m) [inline storage/innobase/buf/buf0flu.cc:2198]: As far as I can tell, the `buf_pool.last_activity_count` could be updated a little earlier, in the only code path where we assigned `idle_flush`. There are no time-consuming steps in between.

**Review by dr-m — COMMENTED** [review]
Thank you. The logic looks OK to me. I only have a few minor suggestions.

File: storage/innobase/buf/buf0flu.cc [hunk 9]
```diff
@@ -2095,9 +2123,14 @@ static os_thread_ret_t DECLARE_THREAD(buf_flush_page_cleaner)(void*)
...
-      my_cond_wait(&buf_pool.do_flush_list,
-                   &buf_pool.flush_list_mutex.m_mutex);
+
+    /* If buf pager cleaner is idle and there is no work
+    (either dirty pages are all flushed or adaptive flushing
+    is not enabled) then opt for non-timed wait */
+    if (buf_pool.page_cleaner_idle() &&
+	(!UT_LIST_GET_LEN(buf_pool.flush_list) ||
+	 srv_max_dirty_pages_pct_lwm == 0.0))
+      my_cond_wait(&buf_pool.do_flush_list, &buf_pool.flush_list_mutex.m_mutex);
```
[the remark this answers is not in the document]
Comment (mysqlonarm) [reply storage/innobase/buf/buf0flu.cc]: done.
my_cond_wait I still see 80 lines. can u re-check.

File: storage/innobase/buf/buf0flu.cc (hunk 7 above)
Comment (mysqlonarm) [reply storage/innobase/buf/buf0flu.cc]: done

File: storage/innobase/buf/buf0flu.cc (hunk 8 above)
Comment (mysqlonarm) [reply storage/innobase/buf/buf0flu.cc]: done
